### DISCUSSION TITLE
setDefaultChain with Web3Modal **V3**

### DISCUSSION THREAD

**Discussion by paouvrard** [WalletConnect/walletconnect-monorepo#3746] [Developer Support]
With Web3Modal **V2**, dApps that implement a chain selector could use `setDefaultChain` before opening the modal which sets the default chain in `requiredNamespaces` of the WalletConnect proposal.
How can I reproduce this behavior with Web3Modal **V3** ? or is there a way to simply not specify requiredNamespaces and have all chains in optionalNamespaces like [here](https://se-sdk-dapp.vercel.app/) ?

**Comment by glitch-txs** [comment]
the first chain of the configuration array will be set as default chain, you can also use [wagmi's useSwitchNetwork hook](https://wagmi.sh/react/hooks/useSwitchNetwork)

You can also use [this option](https://docs.walletconnect.com/web3modal/react/options#defaultchain)

**Reply by paouvrard** [reply]
My question is how to programmatically change the default chain before the wallet is connected (setDefaultChain in V2) which affects `requiredNamespaces`, it's not about switching network or the initial defaultChain configuration.

**Reply by glitch-txs** [reply]
`switchNetwork` will trigger setDefaultChain internally when using a WalletConnect connection

**Reply by paouvrard** [reply]
Calling `switchNetwork` before the wallet is connected didn't have an impact on the default chain, it is still the first chain of the configuration array which is requested in `requiredNamespaces` of the WalletConnect proposal.
`switchNetworkAsyc` throws `Connector not found error` which was expected since the wallet is not yet connected.

**Reply by glitch-txs** [reply]
I'm not sure what's the use case? if you can set it already as [default in config](https://docs.walletconnect.com/web3modal/react/options#defaultchain) or later with [switchChain function](https://wagmi.sh/react/hooks/useSwitchNetwork).

**Reply by paouvrard** [reply]
Some wallets run on only one chain which may not be the one set in the defaultChain config. So the dApp can have a `Connect to Network X` button which will set the default chain to Network X and allow such wallets to connect, otherwise requiredNamespaces will prevent the connection.
An email was sent out by the WalletConnect team on Oct 4:
<img width="404" alt="image" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/29397451/ed3731c4-d50e-4eac-8a20-8b35c970c4c0">
Maybe the team is still working on it but I thought it would be available in V3.
I think V3 should have a way to reproduce these usecases since they were available in V2. Either by modifying the default chain before connecting a wallet, or by having requireNamespaces as empty.

**Reply by glitch-txs** [reply]
Default chain is different from required namespaces, default chain is the chain that will target the transaction while required chain is a chain that MUST be supported for the session to be approved, however I'm not sure if this hook would affect the required chain before connecting in v2, could you confirm this?. 
V3 uses wagmi as well as v2 and they'll set one required chain in the required namespaces and the rest as optional namespaces. Related to the email, we'll soon update the wagmi connector to support only optional namespaces so said configuration won't be required and all wallets are going to be able to connect automatically.

We have a modal view for chains in v3 that you can use to set the required chain before connecting, for this you can use [web components](https://docs.walletconnect.com/web3modal/react/components#w3m-network-button-) or the [open](https://docs.walletconnect.com/web3modal/react/hooks#useweb3modal) function.

Let me know if you are trying to create your own network view, we can add a temporary workaround until we update the Wagmi connector.

**Reply by paouvrard** [reply]
Thank you for the details.

>I'm not sure if this hook would affect the required chain before connecting in v2, could you confirm this?

Switching network before connecting the wallet didn't change the requireNamespaces (still used the first chain of the configuration array)

>V3 uses wagmi as well as v2 and they'll set one required chain in the required namespaces and the rest as optional namespaces.

That's right, and in Web3Modal V2 there is the setDefaultChain hook which enables modifying the requiredNamespaces before the wallet is connected.

>we'll soon update the wagmi connector to support only optional namespaces so said configuration won't be required and all wallets are going to be able to connect automatically

OK, that will solve my problem for using Web3Modal V3.

>Let me know if you are trying to create your own network view, we can add a temporary workaround until we update the Wagmi connector.

I'm using a Connect to Network X button which acts as a network view since the user is selecting the network to connect to so open({ view: 'Network' }) will be redundant. Using Web3Modal V2 is still ok for now, I can switch to V3 when wagmi is updated with only optional namespaces.

**Reply by paouvrard** [reply]
However the wagmi update solution to use only optional namespaces will not work for injected providers right ? So connecting to an injected provider would still require the wallet to support the default network as the first chain of the configuration array. Looks like V3 would greatly benefit from the same `setDefaultChain` hook that was in V2 in order to support use-cases where the network selection is performed within the dApp (not in Web3Modal's network view) and before a wallet is connected.

**Reply by glitch-txs** [reply]
injected ones don't have this issue since they don't work with WalletConnect protocol, we have also integrated EIP-6963 in V3 that improves the UX and functionality of Injected wallets

**Reply by paouvrard** [reply]
ok, also unlike V2, V3 doesn't require to switch to the default chain as part of the wallet connection flow so that's good.
Thanks, looking forward to the wagmi update.
